src/main/java/com/xwiki/authentication/saml/xwiki/XWikiUserManager.java

```diff
@@ -0,0 +1,204 @@
+import static com.xwiki.authentication.saml.samlauth.SamlAuthenticator.PROFILE_PARENT;
+import static java.util.Arrays.asList;
+
+public class XWikiUserManager {
+    public static final EntityReference SAML_XCLASS = new EntityReference("SAMLAuthClass", EntityType.DOCUMENT,
+            new EntityReference(XWiki.SYSTEM_SPACE, EntityType.SPACE));
+    public static final EntityReference USER_XCLASS = PROFILE_PARENT;
+    public static final String SAML_ID_XPROPERTY_NAME = "nameid";
+    private static final Logger LOG = LoggerFactory.getLogger(XWikiUserManager.class);
+    private final SamlAuthConfig authConfig;
+    private final EntityReferenceSerializer<String> compactStringEntityReferenceSerializer;
+    private final DocumentReferenceResolver<String> currentMixedDocumentReferenceResolver;
+
+    public XWikiUserManager(SamlAuthConfig authConfig,
+                            EntityReferenceSerializer<String> compactStringEntityReferenceSerializer,
+                            DocumentReferenceResolver<String> currentMixedDocumentReferenceResolver) {
+        this.authConfig = authConfig;
+        this.compactStringEntityReferenceSerializer = compactStringEntityReferenceSerializer;
+        this.currentMixedDocumentReferenceResolver = currentMixedDocumentReferenceResolver;
+    }
+
+    public DocumentReference getOrCreateUserIfNeeded(XWikiContext context, Saml2XwikiAttributes attributes) throws XWikiException {
+        return new UserManagerWithContextAndAttributes(context, attributes).getOrCreateUserIfNeeded();
+    }
+
+    class UserManagerWithContextAndAttributes {
+        private final XWikiContext context;
+        private final Saml2XwikiAttributes attributes;
+
+        public UserManagerWithContextAndAttributes(XWikiContext context, Saml2XwikiAttributes attributes) {
+            this.context = context;
+            this.attributes = attributes;
+        }
+
+        public DocumentReference getOrCreateUserIfNeeded() throws XWikiException {
+            final User user = getXWikiUsername(attributes.nameID);
+
+            final String wikiId = context.getWikiId();
+            try {
+                // Switch to main wiki to force users to be global users
+                context.setWikiId(context.getMainXWiki());
+
+                if(user.exists())
+                    return syncUserFields(user);
+
+                return createUserWithAttributes(user,attributes);
+            } finally {
+                context.setWikiId(wikiId);
+            }
+        }
+
+        private DocumentReference createUserWithAttributes(User user, Saml2XwikiAttributes attributes) throws XWikiException {
+            LOG.info("Will create new user [{}]", user);
+
+            user.createUserWithAttributes(attributes.xwikiAttributes);
+
+            LOG.info("User [{}] has been successfully created", user);
+
+            return user.getUserReference();
+        }
+
+        private User getXWikiUsername(String nameID)

```

Comment:
Since this returns a `User`, suggest renaming it to 
```suggestion
        private User getXWikiUserByNameID(String nameID)
```